mem_db.go

```diff
@@ -189,25 +158,24 @@ func (db *MemDB) NewBatch() Batch {
 	return &memBatch{db, nil}
 }
 
-//----------------------------------------
-// Iterator
-
 // Implements DB.
 func (db *MemDB) Iterator(start, end []byte) (Iterator, error) {
-	db.mtx.Lock()
-	defer db.mtx.Unlock()
-
+	db.mtx.RLock()
+	defer db.mtx.RUnlock()
 	return newMemDBIterator(db.btree, start, end, false), nil
 }
 
 // Implements DB.
 func (db *MemDB) ReverseIterator(start, end []byte) (Iterator, error) {
-	db.mtx.Lock()
-	defer db.mtx.Unlock()
-
+	db.mtx.RLock()
+	defer db.mtx.RUnlock()
 	return newMemDBIterator(db.btree, start, end, true), nil
 }
 
+//----------------------------------------
+// Iterator
```

Comment:
We've been talking for a long time about getting rid of these ascii separators and instead putting things in their own files. Do you want to do that while you're in here cleaning things up? (Could also go in a follow up PR, since it might create a lot of noise in this one.)

Reply:
Sure thing

Reply:
I split the iterator and batch into separate files, but kept them in the root directory since using a separate directory would break the API quite severely. I also renamed files from e.g. `mem_db_iterator.go` to `memdb_iterator.go`.

Let me know if you think this is fine, and I'll merge this and submit a PR that does the same for the other backends.